sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/ReleaseDownloadListener.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+package com.microsoft.appcenter.distribute;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.Intent;
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.support.annotation.WorkerThread;
+
+import com.microsoft.appcenter.distribute.download.ReleaseDownloader;
+import com.microsoft.appcenter.utils.AppCenterLog;
+import com.microsoft.appcenter.utils.HandlerUtils;
+import com.microsoft.appcenter.utils.storage.SharedPreferencesManager;
+
+import java.text.NumberFormat;
+import java.util.Locale;
+
+import static com.microsoft.appcenter.distribute.DistributeConstants.DOWNLOAD_STATE_ENQUEUED;
+import static com.microsoft.appcenter.distribute.DistributeConstants.HANDLER_TOKEN_CHECK_PROGRESS;
+import static com.microsoft.appcenter.distribute.DistributeConstants.KIBIBYTE_IN_BYTES;
+import static com.microsoft.appcenter.distribute.DistributeConstants.LOG_TAG;
+import static com.microsoft.appcenter.distribute.DistributeConstants.MEBIBYTE_IN_BYTES;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_DISTRIBUTION_GROUP_ID;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_RELEASE_HASH;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_RELEASE_ID;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOAD_STATE;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOAD_TIME;
+import static com.microsoft.appcenter.distribute.InstallerUtils.getInstallIntent;
+
+/**
+ * Listener for downloading progress.
+ */
+class ReleaseDownloadListener implements ReleaseDownloader.Listener {
+
+    /**
+     * Context.
+     */
+    protected Context mContext;
+
+    /**
+     * Last download progress dialog that was shown.
+     * Android 8 deprecates this dialog but only reason is that they want us to use a non modal
+     * progress indicator while we actually use it to be a modal dialog for forced update.
+     * They will always keep this dialog to remain compatible but just mark it deprecated.
+     */
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    private android.app.ProgressDialog mProgressDialog;
+
+    /**
+     * Private field to store information about release we are currently working with.
+     */
+    private ReleaseDetails mReleaseDetails;
```

Comment:
```suggestion
    private final ReleaseDetails mReleaseDetails;
```
This helps understanding thread safety and matches our conventions.